include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h

```diff
@@ -0,0 +1,243 @@
+//===--- UpdatingInstructionIterator.h --------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Classes for tracking instruction iterators to be updated when instructions
+/// are added or deleted.
+///
+/// UpdatingInstructionIteratorRegistry typically resides in the currently
+/// active InstructionDeleter object.
+///
+/// UpdatingInstructionIterator is the iterator adapter.
+/// It is produced by: UpdatingInstructionIteratorRegistry::makeIterator()
+///
+/// Iterators are typically encapsulated in a range returned by
+/// UpdatingInstructionIteratorRegistry::iteratorRange() for use in
+/// range-based for loops:
+///
+///   for (SILInstruction *inst : registry.iteratorRange(bb)) ...
+///
+/// Or more commonly, directly from the InstructionDeleter:
+///
+///   for (SILInstruction *inst : deleter.iteratorRange(bb)) ...
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_UPDATINGINSTRUCTIONITERATOR_H
+#define SWIFT_SILOPTIMIZER_UTILS_UPDATINGINSTRUCTIONITERATOR_H
+
+#include "swift/SIL/SILInstruction.h"
+#include <utility>
+
+namespace swift {
+
+class UpdatingInstructionIteratorRegistry;
+
+namespace detail {
+
+struct UpdatingInstructionIteratorState {
+  using Self = UpdatingInstructionIteratorState;
+
+  SILBasicBlock::iterator base;
```

Comment:
If you make this a template type, it will also work with the reverse_iterator

Reply:
Agreed. We should support reverse_iterator.